mbf_mesh_nav/src/mesh_navigation_server.cpp

```diff
@@ -37,13 +37,94 @@
 
 #include "mbf_mesh_nav/mesh_navigation_server.h"
 
+#include <algorithm>
 #include <functional>
 
 #include <geometry_msgs/msg/pose_array.hpp>
 #include <mesh_map/mesh_map.h>
 #include <nav_msgs/msg/path.hpp>
 #include <rclcpp/logging.hpp>
 
+namespace {
+//! Helper function, intended for formatting available plugin types.
+//! Returns a string like this: [el1, el2, el3, ..., eln]
+std::string stringVectorToString(const std::vector<std::string>& vec) 
+{
+  std::stringstream s;
+  s << "[";
+  bool is_first = true;
+  for (const auto& item : vec)
+  {
+    s << (is_first ? "" : ", ") << item;
+    is_first = false;
+  }
+  s << "]";
+  return s.str();
+};
+
+/*!
+ * Helper function to reduce code duplication when loading plugins.
+ * Things can be either planners, controllers or recovery behaviors.
+ * @tparam AbstractThing Abstract type of the thing to load. Needs to be the base class of SimpleThing and MeshThing. E.g. mbf_abstract_core::AbstractPlanner.
+ * @tparam SimpleThing Simple type of the thing to load. E.g. mbf_simple_core::SimplePlanner
+ * @tparam MeshThing Mesh type of the thing to load. E.g. mbf_simple_core::MeshPlanner
+ * @param thing_type typename of the thing that shall be loaded, usually configured by user via ros params.
+ * @param which_thing should be "planner", "controller", or "recovery behavior", used only for logging.
+ */
+template<typename AbstractThing, typename MeshThing, typename SimpleThing>
+typename AbstractThing::Ptr loadPlugin(const std::string& thing_type, pluginlib::ClassLoader<MeshThing>& mesh_thing_loader, pluginlib::ClassLoader<SimpleThing>& simple_thing_loader, const std::string& which_thing, rclcpp::Logger logger)
```

Comment:
I would suggest placing this templated function into a tcc file under include/mbf_mesh_nav/impl/mesh_navigation_server.tcc or util or something to also allow using it later in mesh_navigation_server derivations (if those will ever exist)

Reply:
I would prefer to keep it in an anonymous namespace, until we really want to re-use the functionality in another context.
This way, it is not accessible from outside the compilation unit and it will be clear to future-me (or you) that the function is only used as a small helper inside here. To me, this implies that it can be easily changed and might not be usable in another context.
Before making it available in a broader context, I'd like to add some more guards against misuse, e.g. ensuring that `AbstractThing` is, in fact, the base class of `MeshThing` and `SimpleThing`. Also, currently, you can e.g. call `loadPlugin(..., simple_loader, mesh_loader, ...)`, which would set `MeshThing=mbf_simple_core::SimplePlanner` and `SimpleThing=mbf_mesh_core::MeshPlanner`, ughh... :face_exhaling: